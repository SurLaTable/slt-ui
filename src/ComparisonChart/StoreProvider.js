import React from 'react';
// import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import sltReducers from './reducers/sltReducers';

const sltStore = createStore(
    sltReducers,
    // Enable DevTools, switch to localStorage on Prod:
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
);

const StoreProvider = (props) => (
    <Provider store={sltStore}>
			<React.Fragment>{props.children}</React.Fragment>
    </Provider>
);
export default StoreProvider;
