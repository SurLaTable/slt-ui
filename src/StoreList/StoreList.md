```jsx {"props":{"style":{"maxHeight":"500px"}}}
<StoreProvider>
	<StoreList sortBy={'alphabetically'} limit={10} detailed={false} />
</StoreProvider>
```

```html
<div data-component="StoreProvider">
	<div data-component="StoreList" data-prop-sort-by="distance"></div>
</div>
```
