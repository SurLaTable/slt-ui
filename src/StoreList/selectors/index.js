import { createSelector } from 'reselect';

import * as sltStoresApi from '../../services/slt-stores';
import {
	filterItemsAlphabetically,
	filterItemsByCulinary
} from '../../services/slt-stores/filters.js';
import * as googleMapsApi from '../../services/google-maps';

function haversineDistance(lat1, lon1, lat2, lon2, unit) {
	var radlat1 = (Math.PI * lat1) / 180;
	var radlat2 = (Math.PI * lat2) / 180;
	var theta = lon1 - lon2;
	var radtheta = (Math.PI * theta) / 180;
	var dist =
		Math.sin(radlat1) * Math.sin(radlat2) +
		Math.cos(radlat1) * Math.cos(radlat2) * Math.cos(radtheta);
	dist = Math.acos(dist);
	dist = (dist * 180) / Math.PI;
	dist = dist * 60 * 1.1515;
	if (unit == 'K') {
		dist = dist * 1.609344;
	}
	if (unit == 'N') {
		dist = dist * 0.8684;
	}
	return dist;
}
function haversineSort(items = [], latlng = {}) {
	return items
		.map((item) => {
			return {
				...item,
				distance: haversineDistance(
					latlng.lat,
					latlng.lng,
					item.location.lat,
					item.location.lng
				)
			};
		})
		.sort((a, b) => a.distance - b.distance);
}
export const getClosestStores = createSelector(
	[sltStoresApi.selectors.getItems, googleMapsApi.selectors.getData],
	(items = [], locationData) => {
		if (locationData?.[0]?.geometry?.location) {
			return haversineSort(items, locationData?.[0]?.geometry?.location);
		} else {
			return filterItemsAlphabetically(items);
		}
	}
);

export const getClosestCulinaryStores = createSelector(getClosestStores, filterItemsByCulinary);
