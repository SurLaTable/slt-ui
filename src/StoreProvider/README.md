```
<div data-component="StoreProvider"></div>
```
