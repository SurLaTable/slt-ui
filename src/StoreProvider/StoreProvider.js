import React from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import configureStore from '../utils/storeTools';

export const store = configureStore();

const StoreProvider = (props) => (
	<Provider store={store}>
		<React.Fragment>{props.children}</React.Fragment>
	</Provider>
);

StoreProvider.propTypes = {
	children: PropTypes.node
};

export default StoreProvider;
