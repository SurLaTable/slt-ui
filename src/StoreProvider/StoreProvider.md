```jsx
<StoreProvider />
```

```html
<div data-component="StoreProvider"></div>
```
