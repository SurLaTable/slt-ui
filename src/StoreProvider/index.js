import React from 'react';
import { Provider } from 'react-redux';
import configureStore from '../utils/storeTools';

export const store = configureStore();
const StoreProvider = (props) => (
	<Provider store={store}>
		<React.Fragment>{props.children}</React.Fragment>
	</Provider>
);
export default StoreProvider;
