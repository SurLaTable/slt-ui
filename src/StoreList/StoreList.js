import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import StoreCard from '../StoreCard';
import List from '@material-ui/core/List';
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';

import * as sltStoresApi from '../services/slt-stores';
import { selectors as googleMapsSelectors } from '../services/google-maps';
import { getClosestStores, getClosestCulinaryStores } from './selectors';

import moize from 'moize';

const styles = (theme) => {
	return {
		button: {
			borderRadius: 0
		},
		StoreList: {
			overflow: 'auto',
			'& > :not(:last-child)': {
				position: 'relative',
				borderBottom: `1px solid ${theme.palette.grey[300]}`
			}
		}
	};
};

class StoreList extends React.Component {
	constructor() {
		super();
		this.state = {
			page: 0
		};
		this.handleShowMore = this.handleShowMore.bind(this);
		this.renderCards = moize.reactSimple(this.renderCards.bind(this));
	}
	handleShowMore() {
		this.setState({
			page: this.state.page + 1
		});
		if (this.props.onShowMore) {
			this.props.onShowMore(this.state.page + 1);
		}
	}
	componentDidMount() {
		if (this.props.hasLocations && this.state.page == 0) {
			this.setState({ page: 1 });
		}
	}
	renderCards({ limit, storeData, detailed, onItemSelected, selectedStore }) {
		let { classes, dispatch } = this.props;
		let cards = [];
		for (let i = 0; i < limit; i++) {
			cards.push(
				<StoreCard
					key={i}
					storeId={storeData[i].storeId}
					name={storeData[i].name}
					location={storeData[i].location}
					contactInfo={detailed ? storeData[i].contact : null}
					hours={detailed ? storeData[i].hours : null}
					distance={storeData[i].distance}
					elevation={0}
				>
					<div
						style={{
							display: 'flex',
							alignItems: 'center',
							justifyContent: 'center',
							height: '100%'
						}}
					>
						{selectedStore != storeData[i].storeId && (
							<Button
								variant="outlined"
								className={classes.button}
								onClick={() => {
									if (selectedStore == storeData[i].storeId) {
										dispatch(sltStoresApi.actions.nullifySelectedItem());
									} else {
										dispatch(
											sltStoresApi.actions.setSelectedItem(
												storeData[i].storeId
											)
										);
										if (onItemSelected) {
											onItemSelected(storeData[i].storeId);
										}
									}
								}}
							>
								Select Location
							</Button>
						)}
					</div>
				</StoreCard>
			);
		}
		return <>{cards}</>;
	}
	componentDidUpdate(prevProps, prevState) {
		if (prevProps.hasLocations != this.props.hasLocations && prevState.page == 0) {
			this.setState({ page: 1 });
		}
	}
	render() {
		//add filters for culinary, and a variable for displaying filters
		let {
			pageSize,
			storeData,
			classes,
			detailed,
			onItemSelected,
			selectedStore,
			hasLocations
		} = this.props;

		let { page } = this.state;

		let limit = Math.min(Number(pageSize) * page, storeData.length);
		let atMax = page > 0 && limit == storeData.length;

		return (
			<>
				<Paper
					elevation={0}
					className={classes.StoreList}
				>
					{this.renderCards({
						limit,
						storeData,
						detailed,
						onItemSelected,
						selectedStore
					})}
				</Paper>
				{atMax === false && (
					<Button
						fullWidth
						onClick={this.handleShowMore}
					>
						Show {page > 0 ? 'More' : 'Stores'}
					</Button>
				)}
			</>
		);
	}
}

StoreList.propTypes = {
	sortBy: PropTypes.string,
	culinary: PropTypes.bool,
	pageSize: PropTypes.number,
	detailed: PropTypes.bool,
	onItemSelected: PropTypes.func,
	onShowMore: PropTypes.func
};
StoreList.defaultProps = {
	sortBy: 'name',
	pageSize: 10,
	detailed: false,
	culinary: false
};

const mapStateToProps = (state, props) => {
	const selectedStore = sltStoresApi.selectors.getSelectedItem(state);
	const hasLocations = googleMapsSelectors.getHasLocations(state);
	let storeData = [];
	if (props.culinary) {
		storeData =
			props.sortBy == 'distance'
				? getClosestCulinaryStores(state)
				: sltStoresApi.selectors.getCulinaryItemsAlphabetically(state);
	} else {
		storeData =
			props.sortBy == 'distance'
				? getClosestStores(state)
				: sltStoresApi.selectors.getItemsAlphabetically(state);
	}

	return {
		...props,
		hasLocations,
		selectedStore,
		storeData
	};
};

export default connect(mapStateToProps)(withStyles(styles)(StoreList));
