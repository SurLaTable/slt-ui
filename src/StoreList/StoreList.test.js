import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import StoreList from './StoreList';
import StoreProvider from '../StoreProvider';

test('Render', () => {
	const renderer = new ShallowRenderer();
	renderer.render(
		<StoreProvider>
			<StoreList />
		</StoreProvider>
	);
	const result = renderer.getRenderOutput();
	expect(result).toMatchSnapshot();
});
