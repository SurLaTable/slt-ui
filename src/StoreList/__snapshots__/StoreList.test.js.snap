// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1`] = `
<Provider
  store={
    Object {
      "asyncReducers": Object {
        "C:\\\\Users\\\\rmcquen\\\\slt-ui\\\\src\\\\services\\\\google-maps": [Function],
        "C:\\\\Users\\\\rmcquen\\\\slt-ui\\\\src\\\\services\\\\slt-stores": [Function],
      },
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <React.Fragment>
    <Connect(WithStyles(StoreList)) />
  </React.Fragment>
</Provider>
`;
